lib/client/doc.js

```diff
@@ -266,33 +266,39 @@ Doc.prototype._emitResponseError = function(err, callback) {
   this.emit('error', err);
 };
 
-Doc.prototype._handleFetch = function(err, snapshot) {
+Doc.prototype._handleFetch = function(error, snapshot) {
+  var callbacks = this.pendingFetch;
```

Comment:
This is something I've done which I _think_ should be okay (and all the tests pass), but I'd like a sense check. The `pendingFetch` callbacks _used_ to be handled by `Doc._resubscribe`, where they'd be queued on the appropriate call.

However, I've sort of just "short-cutted" that behaviour, and now just grab these in both `_handleFetch` and `_handleSubscribe`. It seemed simpler, and I can't seem to think of any downsides...?

If I understand the semantics of this correctly, it's just an array of callbacks that should be called once we receive a snapshot, so grabbing them directly in the response handlers should be the same as trying to smoosh them into the `inflightFetch` callback?

Reply:
The old behaviour was actually that each `fetch` would still trigger a message sent to the server, which would then only call a _single_ callback (`inflightFetch.shift()`), whereas we currently early-return all of them.

We should at the very least mimic the old behaviour (assume that all requests are responded to in order), or — better yet — store callbacks by `seq` like in Presence, and we can invoke the "correct" callback.

Reply:
Per our investigation, old behavior was that, if the connection is online, a reply to fetch will only call the earliest-queued callback in the non-bulk case.

Reply:
@ericyhwang I've thought about it again, and I think that actually the refactored behaviour _is_ still correct.

`pendingFetch` is only ever dealt with when we go offline. Here's how the old behaviour worked:

## Always online

We never touch `pendingFetch`.

## Subscribe whilst offline

- Start offline
- Try to subscribe
- Callback pushed onto `pendingFetch`
- Connection opened
- Call `_resubscribe()`
- Pending subscribe callbacks all combined and used as a single callback for a single subscribe request (or a fetch, if we don't want to be subscribed)

## Lose connection waiting for subscription

- Start online
- Send subscribe request
- Callback pushed onto `inflightSubscribe`
- Connection closed